lib/common/xml.c

```diff
@@ -460,6 +460,73 @@ pcmk__xe_first_child(const xmlNode *parent, const char *node_name,
     return NULL;
 }
 
+/*!
+ * \internal
+ * \brief Set an XML attribute, expanding \c ++ and \c += where appropriate
+ *
+ * If \p target already has an attribute named \p name set to an integer value
+ * and \p value is an addition assignment expression on \p name, then expand
+ * \p value to an integer and set attribute \p name to the expanded value in
+ * \p target.
+ *
+ * Otherwise, set attribute \p name on \p target using the literal \p value.
+ *
+ * For example, suppose \p target has an attribute named \c "X" with value
+ * \c "5", and that \p name is \c "X".
+ * * If \p value is \c "X++", the new value of \c "X" in \p target is \c "6".
+ * * If \p value is \c "X+=3", the new value of \c "X" in \p target is \c "8".
+ * * If \p value is \c "val", the new value of \c "X" in \p target is \c "val".
+ * * If \p value is \c "Y++", the new value of \c "X" in \p target is \c "Y++".
+ *
+ * \param[in,out] target  XML node whose attribute to set
+ * \param[in]     name    Name of the attribute to set
+ * \param[in]     value   New value of attribute to set
+ *
+ * \note The original attribute value in \p target and the number in an
+ *       assignment expression in \p value are parsed as scores. For more
+ *       details, refer to \c char2score().
+ */
+void
+pcmk__xe_set_score(xmlNode *target, const char *name, const char *value)
+{
+    const char *old_value = NULL;
+
+    CRM_CHECK((target != NULL) && (name != NULL) && (value != NULL), return);
+
+    old_value = crm_element_value(target, name);
+
+    // If no previous value, skip to default case and set the value unexpanded.
+    if (old_value != NULL) {
+        const char *n = name;
+        const char *v = value;
+
+        // Stop at first character that differs between name and value
+        for (; (*n == *v) && (*n != '\0'); n++, v++);
```

Comment:
If they're not the same length, then we'll exit when `(*n != *v)` because one of them isn't `'\0'`. The `'\0'` check is really checking for the case when they're both `'\0'`